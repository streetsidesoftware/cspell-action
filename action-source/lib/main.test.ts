import * as process from 'process';
import * as cp from 'child_process';
import * as path from 'path';
import * as fs from 'fs';

const root = path.resolve(path.join(__dirname, '..', '..'));
const fixturesLocation = path.join(root, 'fixtures');
const mainFile = path.join(root, 'action', 'main.js');

describe('Validate Main', () => {
    test('GITHUB_TOKEN', () => {
        expect(process.env['GITHUB_TOKEN']).not.toBeUndefined();
    });

    test('event pull_request main.js', () => {
        const env = fetchGithubActionFixture('pull_request.json');
        const options: cp.ExecSyncOptions = {
            env,
            cwd: root,
        };
        const result = cp.execSync(`node ${mainFile}`, options).toString();

        expect(cleanResult(result)).toMatchSnapshot();
    });

    test('event push main.js', () => {
        const env = fetchGithubActionFixture('push.json');
        const options: cp.ExecSyncOptions = {
            env,
            cwd: root,
        };
        const result = cp.execSync(`node ${mainFile}`, options).toString();
        expect(cleanResult(result)).toMatchSnapshot();
    });
});

function cleanResult(output: string): string {
    // Remove time
    output = output.replace(/\(\d+\.\d\dms\)$/gm, '(???.??ms)');

    const rootRegExp = new RegExp(escapeRegEx(root), 'g');
    output = output.replace(rootRegExp, '.');

    return output;
}

function escapeRegEx(s: string) {
    return s.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
}

function fetchGithubActionFixture(filename: string): Record<string, string> {
    const githubEnv = JSON.parse(fs.readFileSync(path.resolve(fixturesLocation, filename), 'utf-8'));
    if (githubEnv['GITHUB_EVENT_PATH']) {
        githubEnv['GITHUB_EVENT_PATH'] = path.resolve(root, githubEnv['GITHUB_EVENT_PATH']);
    }
    const env = {
        ...process.env,
        ...githubEnv,
    };

    return env;
}
