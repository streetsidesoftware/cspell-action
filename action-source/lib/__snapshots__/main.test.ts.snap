// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Main event pull_request main.js 1`] = `
"cspell-action
Pull Request
1/9 ./.github/workflows/test.yml (???.??ms)
2/9 ./.eslintrc.js (???.??ms)
3/9 ./package.json (???.??ms)
4/9 ./.vscode/launch.json (???.??ms)
5/9 ./fixtures/sampleCode/ts/sample.ts (???.??ms)
6/9 ./action.yaml (???.??ms)
7/9 ./README.md (???.??ms)
8/9 ./cspell.json (???.??ms)
9/9 ./.github/workflows/lint.yml (???.??ms)
Done.
"
`;

exports[`Validate Main event push main.js 1`] = `
"cspell-action
Push
1/1 ./.editorconfig (???.??ms)
Done.
"
`;
